### DISCUSSION TITLE
Creating UI without Json

### DISCUSSION THREAD

**Discussion by Ziboo** [cgerchenhp/UE_TAPython_Plugin_Release#9] [General]
Hello,

I just discovered your plugin because I wanted to create UI in python and quickly found out that it was not possible.
Your plugin looks really nice, but I have some questions.
Most of your examples use a Json file for the main UI, I was wondering:

1) Can I create a UI only in python, without using the Json file ?
2) Can I add dynamic elements to the UI, like widgets in a list in python ?

Thanks a lot

**Comment by cgerchenhp** [comment]
Your questions are valuable. 
 
A1. 
using a JSON file to define the interface can consider as a "serialized result" of the interface. It is much clearer to separate the interface definition from the logic code. Suppose we must write UI code only in Python. I think the below steps are workable. 

1. Keep a valid JSON file with empty contents in the field: "Root" Or use Python to create the placeholder file. Currently, the JSON file on disk is the key to distinguishing the different chameleon tools, so we must have one. 
2. Create a serializable Python class, such as a data class, and use python code to create the interface hierarchy for UI. Then dump it into JSON text, replacing what you did in Step 1.
3. In InitPyCmd, call Step 2 and then use the hot reload interface with this [method](https://www.tacolor.xyz/TipsOfDay/Reload_python_when_launch_Chameleon_Tool.html). 
These steps still allow you to create an interface in Python with one click. 
 
A2. 
When you switch the "comparison mode" in the example ObjectDetailViewer, there are dynamic changes to the interface. 
I set the group widget's visibility with "Collapsed" and "Visible".
In short, a dynamic interface needs pre-defined in the UI and then "Collapsed" with different conditions. 
 
SWidget objects such as SButton(not UButton, that is widge in UMG) have no corresponding object in Python, so Python cannot directly manipulate the Slate interface. This is why we still haven't an official solution for python and Slate UI. 
 
In the dev version of TAPython, I tried to create and add child widgets through child JsonObject, but did not publish them for several reasons: 
- Making UI obscurity 
- Finding controls through "aka" can get confusing, widgets with the same name and widget recycling, etc 
 
I'm also thinking about how to creat a fullly dynamic Slate hierarchy in the future. 

My current thinking is like Jinja2 as a template for HTML might make more sense. However, it's not fully dynamic, either.

Cheer~

**Reply by Ziboo** [reply]
Thank you very much for the prompt response.
It is really a shame that Unreal didn't implemented creation of Slate UI with Python.. :(
Most tools that I would need, definitely consist of having a list of dynamic widget in it ...

**Reply by cgerchenhp** [reply]
Can you provide more details about the specific case? Maybe I can have a try

**Comment by cgerchenhp** [comment]
I convert this issue to discussion. Everyone is welcome to leave a comment.
